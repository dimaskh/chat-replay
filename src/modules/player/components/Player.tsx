// Libs
import React from 'react'

// Components
import Header from 'components/Header'
import { Button, Popup } from 'react-chat-elements'

// Styles
import styled from 'styled-components'

export type Props = {
  isFinished: boolean
  isPlaying: boolean
  clearData: () => { type: string }
  replayRecord: () => { type: string }
}

export class Player extends React.PureComponent<Props> {
  public render() {
    const { isFinished } = this.props

    return (
      <StyledPlayer>
        <Header medium>Player</Header>
        <ButtonContainer>
          <Button onClick={this.startDefaultReplay} text={'Replay Default Record'} />
        </ButtonContainer>
        <Popup text={'Replay is finished'} show={isFinished} footerButtons={this.getPopupButtons()} />
      </StyledPlayer>
    )
  }

  private startDefaultReplay = () => {
    const { isPlaying, replayRecord } = this.props

    if (isPlaying) {
      return
    }

    replayRecord()
  }

  private getPopupButtons = () => {
    const { clearData } = this.props

    const ok = {
      color: 'black',
      text: 'OK',
      onClick: () => clearData(),
    }

    return [ok]
  }
}

export const ButtonContainer = styled.div`
  height: 100px;
  display: flex;
  justify-content: center;
  align-items: center;
`

export const StyledPlayer = styled.div`
  display: flex;
  flex-direction: column;
  width: 20vw;
  height: 100%;
  border-right: 1px solid #616161;
  color: white;
`

export default Player
