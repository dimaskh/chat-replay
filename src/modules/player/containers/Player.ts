// Redux
import { connect } from 'react-redux'
import { bindActionCreators, Dispatch } from 'redux'

// Component
import Player from '../components/Player'

// Actions
import { clearData, replayRecord } from '../actions'

// Selectors
import { isFinished$, isPlaying$ } from '../selectors'

// Types
import { StoreState } from 'store/reducer'

const mapStateToProps = (state: StoreState) => ({
  isFinished: isFinished$(state),
  isPlaying: isPlaying$(state),
})

const mapDispatchToProps = (dispatch: Dispatch) => {
  return bindActionCreators(
    {
      clearData,
      replayRecord,
    },
    dispatch
  )
}

export default connect(mapStateToProps, mapDispatchToProps)(Player as React.ComponentType<any>)
