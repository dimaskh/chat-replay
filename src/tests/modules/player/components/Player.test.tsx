// Libs
import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

// Tested components
import Player from 'modules/player/components/Player'

describe('components/Player', () => {
  let component: any
  let props: any

  beforeEach(() => {
    props = {
      isFinished: false,
      isPlaying: false,
      clearData: jest.fn(),
      replayRecord: jest.fn(),
    }

    component = shallow(<Player {...props} />)
  })

  it('should render correctly', () => {
    expect(shallowToJson(component)).toMatchSnapshot()
  })

  it('should render popup if isFinished true', () => {
    component.setProps({ isFinished: true })

    expect(shallowToJson(component)).toMatchSnapshot()
  })

  it('should handle startDefaultReplay', () => {
    component.instance().startDefaultReplay()

    expect(props.replayRecord).toHaveBeenCalled()
  })

  it('should handle startDefaultReplay if isPlaying true', () => {
    component.setProps({ isPlaying: true })

    expect(props.replayRecord).not.toHaveBeenCalled()
  })
})
