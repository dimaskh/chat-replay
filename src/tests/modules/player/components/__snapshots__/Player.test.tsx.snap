// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Player should render correctly 1`] = `
<styled.div>
  <Header
    medium={true}
  >
    Player
  </Header>
  <styled.div>
    <t
      backgroundColor="#3979aa"
      buttonRef={null}
      className={null}
      color="white"
      disabled={false}
      icon={null}
      onClick={[Function]}
      text="Replay Default Record"
      title={null}
      type={null}
    />
  </styled.div>
  <t
    color="#333"
    footerButtons={
      Array [
        Object {
          "color": "black",
          "onClick": [Function],
          "text": "OK",
        },
      ]
    }
    header={null}
    headerButtons={Array []}
    renderContent={null}
    renderFooter={null}
    renderHeader={null}
    show={false}
    text="Replay is finished"
  />
</styled.div>
`;

exports[`components/Player should render popup if isFinished true 1`] = `
<styled.div>
  <Header
    medium={true}
  >
    Player
  </Header>
  <styled.div>
    <t
      backgroundColor="#3979aa"
      buttonRef={null}
      className={null}
      color="white"
      disabled={false}
      icon={null}
      onClick={[Function]}
      text="Replay Default Record"
      title={null}
      type={null}
    />
  </styled.div>
  <t
    color="#333"
    footerButtons={
      Array [
        Object {
          "color": "black",
          "onClick": [Function],
          "text": "OK",
        },
      ]
    }
    header={null}
    headerButtons={Array []}
    renderContent={null}
    renderFooter={null}
    renderHeader={null}
    show={true}
    text="Replay is finished"
  />
</styled.div>
`;
